### ISSUE TITLE
terraform apply returning successful exit code on error sometimes

### ISSUE THREAD

**Issue by automatecloud** [hashicorp/terraform#20671]
### Current Terraform Version
0.11.13

### Use-cases
I'm using terraform inside a bash script to deploy a complex demo environment. After terraform apply other script steps are executed, but it should exit if there are any errors with terraform apply. If there is an error with the deployment there is no change on the exit code created by apply it always will return 0.

I need a feature similar to the terraform plan --detailed-exitcode, to return a detailed exit code when the command exits. When provided, this argument changes the exit codes and their meanings to provide more granular information about what the resulting plan contains:

0 = Succeeded apply
1 = Error

### Attempted Solutions
Right now i'm using the workaround to check for any "Error applying plan" inside the output:
terraform apply | tee  /dev/tty | ( ! grep "Error applying plan" )

### Proposal
An optional parameter --detailed-exitcode for terraform apply would be nice.

### References
https://devops.stackexchange.com/questions/871/terraform-apply-exit-code-on-error
- #13598
- #11302

-->

**Comment by apparentlymart** [comment]
Hi @automatecloud,

The intended behavior is to return a Nonzero exit code on error by default, and we have tests verifying that behavior for several situations but it seems like you've found a case where it isn't working. That's a bug, not a missing feature.

Can you share some more details on exactly how you are running Terraform (the automation code you have around it, how the plan and apply are connected, etc) and then we can try to figure out what is different in your case from the existing test cases?

Thanks!

**Comment by automatecloud** [comment]
The challenge i face is if that sometime the resource is already available. I'm managing DNS entries in GCP and sometimes they were already added by someone manually. If you use terraform apply it exits with a "Error applying plan". My script should stop there as i can't guarantee the DNS entry is configured with the right cname for example. If it doesn't the rest of it could fail. So i'm using the workaround to grep the logs for the "Error applying plan" entry. That is a good workaround, but terraform apply should be able to return the right exit code if i need it.

terraform refresh and terraform plan didn't help as they do not find the GCP resources.

**Comment by apparentlymart** [comment]
Hi @automatecloud,

The thing that is strange to me is that Terraform should _already_ be generating a nonzero exit code in that situation, but apparently in your situation it is not. I'd like to understand more about what you're doing so I can understand how it differs from the various automated tests we have in Terraform that are verifying that errors during `terraform apply` produce a non-zero exit code.

In other words, Terraform should already be behaving the way you describe, so I'd like to try to understand why it isn't in your case.

**Comment by automatecloud** [comment]
It's easy to reproduce. You only need to create a resource manually. Let's say you plan to manage a DNS record. Add them manually inside GCP.  After that defining the resource in a terraform file:

resource "google_dns_record_set" "master" {
  name = "example.andreas.com"
  provider = "google"
  project     = "myproject"
  type = "A"
  ttl  = 60

  managed_zone = "andreas-lab"

  rrdatas = ["10.10.10.10"]
}

and apply it. It will show you the following error

Error applying plan:

1 error(s) occurred:

* google_dns_record_set.master: 1 error(s) occurred:

* google_dns_record_set.master: Error creating DNS RecordSet: googleapi: Error 409: The resource 'entity.change.additions[0]' named 'example.andreas.com. (A)' already exists, alreadyExists

The exit code of terraform is 0 and not a non-zero.

If a terraform apply is not successful it should be always non-zero. 

You can also find another example of that failed to apply because of existing resources here https://github.com/hashicorp/terraform/issues/20344

**Comment by devblackops** [comment]
I am seeing similar behavior with Azure Key Vault.

I run a `terraform plan -out terraform.plan` and see 12 resources to be created.

I then run `terraform apply -auto-approve terraform.plan` and error out creating the Key vault with: 

```
Error: Error applying plan:

1 error(s) occurred:

* azurerm_key_vault.client-keyvault: 1 error(s) occurred:

* azurerm_key_vault.client-keyvault: Error updating Key Vault "xxxxx-keyvault" (Resource Group "tf_int_201949012204_4e11"): keyvault.VaultsClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="BadRequest" Message="An invalid value was provided for 'accessPolicies'."

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

The actual cause for the error is a problem on my end with a bad variable but TF returns a `0` exit code during the apply so this isn't being caught in CI.

IMHO **any** error during the apply should return a non-zero exit code.

**Comment by hendrixroa** [comment]
for me, in the version 0.11.13 the param `detailed-exitcode` doesn't work.
the commands `init`, and `apply` should return non zero exit code.

**Comment by hendrixroa** [comment]
Sorry, my fault, plan and apply return exit code as expected. but would be great if init can have one

**Comment by james-flynn-ie** [comment]
@apparentlymart, I'd like to take this one on.

Have you any advice on where to get started, please?

**Comment by tjj9020** [comment]
Not sure if this is still an issue or not.  Tried to replicate using AWS route53 by creating a record that already exists and the exit code return 1.  Maybe is specific to a provider?

**Comment by mattlord** [comment]
We've also experienced this with the `vault_database_secret_backend_role` resources (e.g. when the database name/schema is a reserved word or the password used is wrong). This is fairly major pain because the problems are not detected in the CI/CD pipeline.

@apparentlymart, any way that I can help?

**Comment by j4g4f** [comment]
Hello,

Can we get an update on this issue?  This continues to be a problem 10 months on, and is making integration with CI/CD more difficult than necessary.

Thank you!

**Comment by bougyman** [comment]
Having the same problem here, exiting 0 when it fails to create a necessary S3 bucket due to the infamous 
```
Error: Error creating S3 bucket: Error creating S3 bucket sp-terraform-s3-state-test
 retrying: OperationAborted: A conflicting conditional operation is currently in progress against this resource. Please try again.
        status code: 409, request id: E3ACB8F77041E863, host id: qCXGnd/ZkZ7ymm25s3v8fiyABmhrmLcLoOGpeNTJ57pgLESQhljeKNHoI93gW8DjI78l5QcgkEk=

  on ../module/s3.tf line 1, in resource "aws_s3_bucket" "s3_terraform_state":                                                         
   1: resource "aws_s3_bucket" "s3_terraform_state" {                                                                                  

                                                                                                                                       
Releasing state lock. This may take a few moments...                                                                                   
me@myhost ~/ws/terraform_sre/Modules/state_management/tf_state_test                                         
 % echo $?                                                                                                                             
0                                                                                           
```

**Comment by tg12** [comment]
Same issue here.

**Comment by ngodec** [comment]
Same issue here - plus, sometimes a successful run returns a non-zero exit code

**Comment by fluffy-cakes** [comment]
Adding my 2 cents that it's happening for me on Azure resources for v.12.20

**Comment by jackmahoney** [comment]
Same problem in context of ZSH and Makefile, running terraform v.12.20. See big red errors but my make task completes successfully.

**Comment by daweins** [comment]
Same issue found in AzureRM with error results from existing diagnostic log settings against network security groups

**Comment by lukaszraczylo** [comment]
Issue still present in 0.13 with both AWS ( insufficient permission ) and PagerDuty provider ( subscription level errors )

**Comment by danieldreier** [comment]
hey folks - I see a lot of people saying that they're encountering similar successful exit codes on errors. I'd like to explain what help we need in order to move this forward. The practical thing I need to translate this into action is very concrete examples, e.g:

1. run the following command(s) to set up the environment
2. run linked terraform config using "terraform apply" on X.Y.Z terraform version
3. output you're seeing

To work on or even prioritize this for engineers, I have to be able to run each case on my workstation without inventing any details in order to be confident we're seeing the same behavior. As-is, it's not clear to me whether these are all the same issue, or if there are multiple different issues (e.g. bugs in different providers), and so without details on reproductions, I'm stuck and can't plan a fix.

In order to figure out whether these are the same or different issues, can you please write up reproduction cases such that I can copy-paste it and run them locally? Ideally, this would use the [null resource](https://www.terraform.io/docs/providers/null/resource.html) provider rather than a real provider in order to minimize external dependencies, but I understand that some of these cases will require real cloud providers. The key is that I need to be able to copy-paste and run this, so the easier to set up, the easier it will be for me to reproduce and prioritize.

**Comment by 123BLiN** [comment]
Any news on this? maybe something was improved in 1.0.0?

**Comment by andrewphaine** [comment]
I have the same issue when trying to create a VM using a release pipeline. Terraform is not able to locate the state file (State read error), yet it does not return an error code, instead, the pipeline completes successfully.

**Comment by andrewphaine** [comment]
> 
> 
> I have the same issue when trying to create a VM using a release pipeline. Terraform is not able to locate the state file (State read error), yet it does not return an error code, instead, the pipeline completes successfully.

Just an update: I updated the Terraform install task to install the latest version - 1.0.3
- I also updated the azurerm provider to the latest version - 2.70.0 - and I did get a failure in the pipeline for Terraform not being able to read secrets from the KeyVault, which was obviously due to the Service Connection not having bee granted access to the KeyVault. So, it does seem to be working correctly now.

**Comment by andrewphaine** [comment]
> 
> 
> Any news on this? maybe something was improved in 1.0.0?

Check 1.0.3 - does seem to be sorted now

**Comment by vladkasianenko** [comment]
Still not working:
![image](https://user-images.githubusercontent.com/14053728/131982421-7eaa1779-4c58-412b-8c90-caf9d43ee6f7.png)

```
2021-09-03T08:51:40.3828050Z Task         : Bash
2021-09-03T08:51:40.3828329Z Description  : Run a Bash script on macOS, Linux, or Windows
2021-09-03T08:51:40.3828595Z Version      : 3.189.0
2021-09-03T08:51:40.3828848Z Author       : Microsoft Corporation
2021-09-03T08:51:40.3829190Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
2021-09-03T08:51:40.3829592Z ==============================================================================
```

Terraform version 1.0.5

**Comment by apparentlymart** [comment]
Hi all,

As far as I can tell from reading the history here, there still isn't an example we can try in our local development environments in order to observe the problems being reported. While I would like to find and fix this problem, we can't do so if we can't observe it happening.

Given the age of this issue and that it was originally opened against a now-obsolete version of Terraform CLI, I'm going to close this issue. However, I don't intend that to mean that we don't want to find and fix the issue, but rather that in order to do so we'd need to see a new issue that includes all of the information requested in our current issue template, and ideally not requiring particular cloud system credentials in order to reproduce, although I understand that might not be possible given that local-only resource types tend not to generate errors like this. If you can only reproduce it with remote objects, please reduce your configuration to as small as possible while still exhibiting the problem, so that we can successfully apply it in a test account and try to observe the result you saw.

---

Please note that if you are running Terraform in a wrapper shell script then typically your shell script must propagate an unsuccessful exit status through to its caller, which is not the default behavior. If you are using Bash then you can tell it to treat any error in an intermediate command as a failure of the overall script by including the following line prior to running Terraform:

```
set -e
```

Alternatively, you can write logic to explicitly check the exit status and use the `exit` command to propagate an unsuccessful code to the shell script's result too.

In order to find and fix this bug we will need to see an example that we can run directly in a terminal using `terraform apply`, without any wrapper scripts or remote automation.

---

Thanks!
